config/testconfig.go

```diff
@@ -39,3 +48,232 @@ type TestConfig struct {
+	)
+	if err != nil {
+		return err
+	}
+	err = c.readSingleEnvVar(
+		"ChainlinkImage.Image",
+		"TEST_CONFIG_CHAINLINK_IMAGE",
+		EnvValueType(String),
+	)
+	if err != nil {
+		return err
+	}
+	err = c.readSingleEnvVar(
+		"ChainlinkUpgradeImage.Image",
+		"TEST_CONFIG_CHAINLINK_UPGRADE_IMAGE",
+		EnvValueType(String),
+	)
+	if err != nil {
+		return err
+	}
+	err = c.readSingleEnvVar(
+		"Logging.Loki.TenantId",
+		"TEST_CONFIG_LOKI_TENANT_ID",
+		EnvValueType(String),
+	)
+	if err != nil {
+		return err
+	}
+	err = c.readSingleEnvVar(
+		"Logging.Loki.Endpoint",
+		"TEST_CONFIG_LOKI_ENDPOINT",
+		EnvValueType(String),
+	)
+	if err != nil {
+		return err
+	}
+	err = c.readSingleEnvVar(
+		"Logging.Loki.BasicAuth",
+		"TEST_CONFIG_LOKI_BASIC_AUTH",
+		EnvValueType(String),
+	)
+	if err != nil {
+		return err
+	}
+	err = c.readSingleEnvVar(
+		"Logging.Loki.BearerToken",
+		"TEST_CONFIG_LOKI_BEARER_TOKEN",
+		EnvValueType(String),
+	)
+	if err != nil {
+		return err
+	}
+	err = c.readSingleEnvVar(
+		"Logging.Grafana.BaseUrl",
+		"TEST_CONFIG_GRAFANA_BASE_URL",
+		EnvValueType(String),
+	)
+	if err != nil {
+		return err
+	}
+	err = c.readSingleEnvVar(
+		"Logging.Grafana.DashboardUrl",
+		"TEST_CONFIG_GRAFANA_DASHBOARD_URL",
+		EnvValueType(String),
+	)
+	if err != nil {
+		return err
+	}
+	err = c.readSingleEnvVar(
+		"Logging.Grafana.BearerToken",
+		"TEST_CONFIG_GRAFANA_BEARER_TOKEN",
+		EnvValueType(String),
+	)
+	if err != nil {
+		return err
+	}
+	err = c.readSingleEnvVar(
+		"Pyroscope.ServerUrl",
+		"TEST_CONFIG_PYROSCOPE_SERVER_URL",
+		EnvValueType(String),
+	)
+	if err != nil {
+		return err
+	}
+	err = c.readSingleEnvVar(
+		"Pyroscope.Key",
+		"TEST_CONFIG_PYROSCOPE_KEY",
+		EnvValueType(String),
+	)
+	if err != nil {
+		return err
+	}
+	err = c.readSingleEnvVar(
+		"Pyroscope.Environment",
+		"TEST_CONFIG_PYROSCOPE_ENVIRONMENT",
+		EnvValueType(String),
+	)
+	if err != nil {
+		return err
+	}
+	err = c.readSingleEnvVar(
+		"Pyroscope.Enabled",
+		"TEST_CONFIG_PYROSCOPE_ENABLED",
+		EnvValueType(Boolean),
+	)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Read env vars for map[string][]string
+func (c *TestConfig) readEnvVarGroups(key string, regexpStr string) error {
+	logger := logging.GetTestLogger(nil)
+	re := regexp.MustCompile(regexpStr)
+
+	envVars := make(map[string][]string)
+	for _, env := range os.Environ() {
+		pair := strings.SplitN(env, "=", 2)
+		if len(pair) != 2 {
+			continue
+		}
+		key, value := pair[0], pair[1]
+
+		matches := re.FindStringSubmatch(key)
+		if matches != nil {
+			key := matches[1]
+			envVars[key] = append(envVars[key], value)
+		}
+	}
+
+	if len(envVars) == 0 {
+		logger.Debug().Msgf("Not setting test config key '%s' because environment variables that match '%s' regex not found", key, regexpStr)
+		return nil
+	}
+
+	for network, keys := range envVars {
+		keyPath := fmt.Sprintf("%s.%s", key, network)
+		viperLock.Lock()
+		viper.Set(keyPath, keys)
+		viperLock.Unlock()
```

Comment:
why do we need to use mutex here? in what situation would we experience concurrent operations on the same test config?

Reply:
@Tofel `viper` does not handle synchonization well. There are multiple issues about it. I just pushed another commit replacing viper - we don't need this new module